modules/geo-layers/src/tile-3d-layer/get-frame-state.js

```diff
@@ -0,0 +1,95 @@
+import {Vector3} from 'math.gl';
+import {CullingVolume, Plane} from '@math.gl/culling';
+import {Ellipsoid} from '@math.gl/geospatial';
+
+const scratchPlane = new Plane();
+const scratchPosition = new Vector3();
+const cullingVolume = new CullingVolume([
+  new Plane(),
+  new Plane(),
+  new Plane(),
+  new Plane(),
+  new Plane(),
+  new Plane()
+]);
+
+// Extracts a frame state appropriate for tile culling from a deck.gl viewport
+// TODO - this could likely be generalized and merged back into deck.gl for other culling scenarios
+export function getFrameState(viewport, timeline) {
+  // Traverse and and request. Update _selectedTiles so that we know what to render.
+  const tick = timeline.getTime();
+  const {cameraDirection, cameraUp, height} = viewport;
+  const {metersPerPixel} = viewport.distanceScales;
+
+  const viewportCenterCartographic = [viewport.longitude, viewport.latitude, 0];
+  // TODO - Ellipsoid.eastNorthUpToFixedFrame() breaks on raw array, create a Vector.
+  // TODO - Ellipsoid.eastNorthUpToFixedFrame() takes a cartesian, is that intuitive?
+  const viewportCenterCartesian = Ellipsoid.WGS84.cartographicToCartesian(
+    viewportCenterCartographic,
+    new Vector3()
+  );
+  const enuToFixedTransform = Ellipsoid.WGS84.eastNorthUpToFixedFrame(viewportCenterCartesian);
+
+  const cameraPositionCartographic = viewport.unprojectPosition(viewport.cameraPosition);
+  const cameraPositionCartesian = Ellipsoid.WGS84.cartographicToCartesian(
+    cameraPositionCartographic,
+    new Vector3()
+  );
+
+  // These should still be normalized as the transform has scale 1 (goes from meters to meters)
+  const cameraDirectionCartesian = new Vector3(
+    enuToFixedTransform.transformAsVector(new Vector3(cameraDirection).scale(metersPerPixel))
+  ).normalize();
+  const cameraUpCartesian = new Vector3(
+    enuToFixedTransform.transformAsVector(new Vector3(cameraUp).scale(metersPerPixel))
+  ).normalize();
+
+  commonSpacePlanesToWGS84(viewport);
+
+  // TODO: make a file/class for frameState and document what needs to be attached to this so that traversal can function
+  return {
+    camera: {
+      position: cameraPositionCartesian,
+      direction: cameraDirectionCartesian,
+      up: cameraUpCartesian
+    },
+    height,
+    cullingVolume,
+    frameNumber: tick, // TODO: This can be the same between updates, what number is unique for between updates?
```

Comment:
Why is `tick` used here? If the user pauses transitions/animations, would it also block tiles from loading?

Reply:
> Why is tick used here? If the user pauses transitions/animations, would it also block tiles from loading?

The traversal code wants a monotonous counter that increases with each frame. This is unrelated to animations. if we are now doing clever things with the `tick`, we could consider adding a new value to the `animationProps`.